### PR TITLE
Sticky queues sample

### PR DESCRIPTION
**Pull request by jakejscott**
## What was changed
Ported the Sticky Activity Queues example from typescript

## Why?
So that we know how to use sticky queues.

## Checklist

1. Closes https://github.com/temporalio/samples-dotnet/issues/13

2. How was this tested:
Manually by running locally

3. Any docs updates needed?
Updated the readmes

### CODE REVIEW

File: src/ActivityStickyQueues/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+- It doesn't matter where this activity is run, so it can be "non sticky" as per Temporal default behavior.
+- In this demo, `uniqueWorkerTaskQueue` is simply a `uuid` initialized in the Worker, but you can inject smart logic here to uniquely identify the Worker, [as Netflix did](https://community.temporal.io/t/using-dynamic-task-queues-for-traffic-routing/3045).
+- For activities intended to be "sticky", only register them in one Worker, and have that be the only Worker listening on that `uniqueWorkerTaskQueue`.
+- Execute workflows from the Client like normal.
+
+Activities have been artificially slowed with `await Task.Delay(TimeSpan.FromSeconds(3))` to simulate slow activities.
+
+### Running this sample
+
+1. `temporal server start-dev` to start [Temporal Server](https://github.com/temporalio/cli/#installation).
```
Comment (cretz) [inline src/ActivityStickyQueues/README.md:17]: Can you link to the base README like the other examples' READMEs do wrt prerequisites?

File: src/ActivityStickyQueues/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+```bash
+Running worker
+[19:17:58] info: Temporalio.Activity:DownloadFileToWorkerFileSystem[0]
+      Downloading https://temporal.io and saving to path C:\Users\jaken\AppData\Local\Temp\tmp4409.tmp
+[19:18:04] info: Temporalio.Activity:WorkOnFileInWorkerFileSystem[0]
+      Did some work on C:\Users\jaken\AppData\Local\Temp\tmp4409.tmp, checksum: b3fc767460efa514753a75e6f3d7af97
+[19:18:07] info: Temporalio.Activity:CleanupFileFromWorkerFileSystem[0]
+      Removing C:\Users\jaken\AppData\Local\Temp\tmp4409.tmp
+```
```
Comment (cretz) [inline src/ActivityStickyQueues/README.md:31]: No need to show example output unless you really want to

File: src/ActivityStickyQueues/FileProcessingWorkflow.cs
Comment (cretz) [inline src/ActivityStickyQueues/FileProcessingWorkflow.cs:1]: We've been putting this in a `.workflow.cs` file

File: src/ActivityStickyQueues/FileProcessingWorkflow.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            try
+            {
+                var uniqueWorkerTaskQueue = await Workflow.ExecuteActivityAsync(
+                    (NonStickyActivities act) => act.GetUniqueTaskQueueAsync(),
+                    new()
+                    {
+                        StartToCloseTimeout = TimeSpan.FromMinutes(1),
+                    });
+
+                var downloadPath = Path.GetTempFileName();
```
Comment (cretz) [inline src/ActivityStickyQueues/FileProcessingWorkflow.cs:23]: This is non-deterministic. Please move this to the activity (i.e. maybe the first activity needs to return this)

File: src/ActivityStickyQueues/FileProcessingWorkflow.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+[Workflow]
+public class FileProcessingWorkflow
+{
+    [WorkflowRun]
+    public async Task RunAsync(int maxAttempts = 5)
+    {
+        for (var attempt = 1; attempt <= maxAttempts; ++attempt)
+        {
+            try
+            {
```
Comment (cretz) [inline src/ActivityStickyQueues/FileProcessingWorkflow.cs:15]: I know TypeScript does it (and arguably shouldn't), but I wouldn't do a for loop and try/catch for attempts. I would let the workflow fail, and if you wanted retry and attempts and such, I would set the retry policy workflow side.

File: src/ActivityStickyQueues/FileProcessingWorkflow.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                    (NonStickyActivities act) => act.GetUniqueTaskQueueAsync(),
+                    new()
+                    {
+                        StartToCloseTimeout = TimeSpan.FromMinutes(1),
+                    });
+
+                var downloadPath = Path.GetTempFileName();
+
+                await Workflow.ExecuteActivityAsync(
+                    (StickyActivities act) => act.DownloadFileToWorkerFileSystemAsync("https://temporal.io", downloadPath),
```
Comment (cretz) [inline src/ActivityStickyQueues/FileProcessingWorkflow.cs:26]: Usually we recommend a `record` for two or more parameters (or even single ones sometimes too) for future proofing when you want to add fields

File: src/ActivityStickyQueues/FileProcessingWorkflow.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                    new()
+                    {
+                        TaskQueue = uniqueWorkerTaskQueue,
+                        // Note the use of ScheduleToCloseTimeout.
+                        // The reason this timeout type is used is because this task queue is unique
+                        // to a single worker. When that worker goes away, there won't be a way for these
+                        // activities to progress.
+                        ScheduleToCloseTimeout = TimeSpan.FromMinutes(1),
```
Comment (cretz) [inline src/ActivityStickyQueues/FileProcessingWorkflow.cs:34]: I know TypeScript doesn't, and arguably should, but for best practices, we like to strongly encourage a `HeartbeatTimeout`

File: src/ActivityStickyQueues/FileProcessingWorkflow.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+                    });
+                try
+                {
+                    await Workflow.ExecuteActivityAsync(
+                        (StickyActivities act) => act.WorkOnFileInWorkerFileSystemAsync(downloadPath),
+                        new()
+                        {
+                            TaskQueue = uniqueWorkerTaskQueue,
+                            ScheduleToCloseTimeout = TimeSpan.FromMinutes(1),
```
Comment (cretz) [inline src/ActivityStickyQueues/FileProcessingWorkflow.cs:43]: Probably want more time to "work" on a file

File: src/ActivityStickyQueues/Program.cs [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+    var nonStickyActivities = new NonStickyActivities(uniqueWorkerTaskQueue);
+    var stickyActivities = new StickyActivities();
+
+    // Run worker until cancelled
+    Console.WriteLine("Running worker");
+
+    using var nonStickyWorker = new TemporalWorker(
+        client,
+        new TemporalWorkerOptions(taskQueue: "sticky-activity-tutorial")
```
Comment (cretz) [inline src/ActivityStickyQueues/Program.cs:35]:
```suggestion
        new TemporalWorkerOptions(taskQueue: "activity-sticky-queues-sample")
```

Matches our other examples better

File: src/ActivityStickyQueues/StickyActivities.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
+using System.Security.Cryptography;
+using System.Text;
+using Microsoft.Extensions.Logging;
+using Temporalio.Activities;
+
+namespace TemporalioSamples.ActivityStickyQueues;
+
+public class StickyActivities
```
Comment (cretz) [inline src/ActivityStickyQueues/StickyActivities.cs:8]: Probably want to make this a static class w/ static methods if there is no state

File: src/ActivityStickyQueues/NonStickyActivities.cs [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+    public NonStickyActivities(string uniqueWorkerTaskQueue)
+    {
+        this.uniqueWorkerTaskQueue = uniqueWorkerTaskQueue;
+    }
+
+    [Activity]
+    public Task<string> GetUniqueTaskQueueAsync()
+    {
+        return Task.FromResult(uniqueWorkerTaskQueue);
+    }
```
Comment (cretz) [inline src/ActivityStickyQueues/NonStickyActivities.cs:18]:
```suggestion
    public NonStickyActivities(string uniqueWorkerTaskQueue) =>
        this.uniqueWorkerTaskQueue = uniqueWorkerTaskQueue;

    [Activity]
    public string GetUniqueTaskQueue() => uniqueWorkerTaskQueue;
```

Not required, just may be easier to read (if .NET gives you guff about getters, you can disable that warning via pragma)

File: src/ActivityStickyQueues/FileProcessingWorkflow.cs (hunk 3 above)
Comment (jakejscott) [reply src/ActivityStickyQueues/FileProcessingWorkflow.cs:23]: Oh yes good catch

Comment (jakejscott) [reply src/ActivityStickyQueues/FileProcessingWorkflow.cs]: I updated it to this: `var downloadPath = Path.Join(Path.GetTempPath(), Workflow.NewGuid().ToString());` 

`Path.GetTempPath` should be deterministic right?

Comment (jakejscott) [reply src/ActivityStickyQueues/FileProcessingWorkflow.cs]: The docs say: "Returns the path of the current user's temporary folder."

Comment (cretz) [reply src/ActivityStickyQueues/FileProcessingWorkflow.cs]: That's not deterministic if it is current-user (or even machine) specific. You should accept this as a response from the first activity and let that activity choose to tell you where they downloaded the file. This also is better practice anyways to let the downloaded choose the temporary location for the file.

File: src/ActivityStickyQueues/Program.cs [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    using var stickyWorker = new TemporalWorker(
+        client,
+        new TemporalWorkerOptions(taskQueue: uniqueWorkerTaskQueue)
+            .AddActivity(StickyActivities.DownloadFileToWorkerFileSystemAsync)
+            .AddActivity(StickyActivities.CleanupFileFromWorkerFileSystemAsync)
+            .AddActivity(StickyActivities.WorkOnFileInWorkerFileSystemAsync));
+
+    try
+    {
+        await Task.WhenAll(nonStickyWorker.ExecuteAsync(tokenSource.Token), stickyWorker.ExecuteAsync(tokenSource.Token));
```
Comment (cretz) [inline src/ActivityStickyQueues/Program.cs:47]: The trouble with this is what if one worker threw a fatal exception not related to cancellation? Now it's just hung waiting on the other properly-working worker. I think this is fine for the sample, but something to think about.

File: src/ActivityStickyQueues/StickyActivities.cs [hunk 12]
```diff
@@ -0,0 +1,50 @@
+using System.Globalization;
+using System.Security.Cryptography;
+using System.Text;
+using Microsoft.Extensions.Logging;
+using Temporalio.Activities;
+
+namespace TemporalioSamples.ActivityStickyQueues;
+
+public record DownloadFileArgs(Uri Uri, string Path);
```
[the remark this answers is not in the document]
Comment (cretz) [reply src/ActivityStickyQueues/StickyActivities.cs:9]: That's the problem with these analyzers is that they don't understand nuance :-)  I'm actually kinda of curious how `Uri` looks serialized to JSON and I assume it works properly when deserialized. But yeah, can disable pragma or change the property name.

File: src/ActivityStickyQueues/Program.cs (hunk 11 above)
Comment (jakejscott) [reply src/ActivityStickyQueues/Program.cs:47]: Curious how to handle this properly... Would you use `Task.WhenAny`?  Closest docs I can find is this https://learn.microsoft.com/en-us/dotnet/csharp/asynchronous-programming/cancel-an-async-task-or-a-list-of-tasks

Comment (cretz) [reply src/ActivityStickyQueues/Program.cs:47]: It gets confusing (and it's not Temporal specific). It'd basically be a when any w/ captured non-cancel exception followed by a shutdown of each (or cancel of token) followed by a when all (maybe w/ exceptions ignored) followed by rethrowing what happened in when any.

Comment (jakejscott) [reply src/ActivityStickyQueues/Program.cs:47]: I tried a couple things to try and implement this, but TBH I'm struggling :)  I'm not sure how to cause either of the workers to blow up. 

This is what I tried: 

Note I added an `otherWorker` which throws after 10 seconds to test this out:

```csharp
async Task RunWorkerAsync()
{
    // Cancellation token cancelled on ctrl+c
    using var tokenSource = new CancellationTokenSource();
    Console.CancelKeyPress += (_, eventArgs) =>
    {
        tokenSource.Cancel();
        eventArgs.Cancel = true;
    };

    var uniqueWorkerTaskQueue = Guid.NewGuid().ToString();

    var nonStickyActivities = new NonStickyActivities(uniqueWorkerTaskQueue);

    // Run worker until cancelled
    Console.WriteLine("Running worker");

    using var nonStickyWorker = new TemporalWorker(
        client,
        new TemporalWorkerOptions(taskQueue: "activity-sticky-queues-sample")
            .AddActivity(nonStickyActivities.GetUniqueTaskQueue)
            .AddWorkflow<FileProcessingWorkflow>());

    using var stickyWorker = new TemporalWorker(
        client,
        new TemporalWorkerOptions(taskQueue: uniqueWorkerTaskQueue)
            .AddActivity(StickyActivities.DownloadFileToWorkerFileSystemAsync)
            .AddActivity(StickyActivities.CleanupFileFromWorkerFileSystemAsync)
            .AddActivity(StickyActivities.WorkOnFileInWorkerFileSystemAsync));

    using var otherWorker = new TemporalWorker(
        client,
        new TemporalWorkerOptions(taskQueue: "blah").AddActivity(nonStickyActivities.GetUniqueTaskQueue));

    var nonStickyWorkerTask = nonStickyWorker.ExecuteAsync(tokenSource.Token);
    var stickyWorkerTask = stickyWorker.ExecuteAsync(tokenSource.Token);
    var otherWorkerTask = otherWorker.ExecuteAsync(
        async () =>
        {
            await Task.Delay(TimeSpan.FromSeconds(10));
            throw new ArgumentException("Something bad happened");
        },
        tokenSource.Token);

    var task = await Task.WhenAny(nonStickyWorkerTask, stickyWorkerTask, otherWorkerTask);
    if (task.Exception is not null)
    {
        try
        {
            tokenSource.Cancel();
            await Task.WhenAll(nonStickyWorkerTask, stickyWorkerTask, otherWorkerTask);
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Unhandled ex {ex}");
            throw;
        }
    }
    else
    {
        Console.WriteLine("Worker cancelled");
    }
}
```

I the logs I see a warning `Worker finalization failed`

```
Running worker
[13:07:15] info: Temporalio.Activity:DownloadFileToWorkerFileSystem[0]
      Downloading https://temporal.io and saving to path C:\Users\jaken\AppData\Local\Temp\tmp92FC.tmp
[13:07:21] info: Temporalio.Activity:WorkOnFileInWorkerFileSystem[0]
      Did some work on C:\Users\jaken\AppData\Local\Temp\tmp92FC.tmp, checksum: 49d7419e6cba3575b3158f62d053f922aa08b23c64f05411cda3213b56c84ba4   
[13:07:22] info: Temporalio.Worker.TemporalWorker[0]
      User task completed, shutting down
[13:07:22] info: Temporalio.Worker.ActivityWorker[0]
      Beginning activity worker shutdown, will wait 00:00:00 before cancelling 0 activity instance(s)
[13:07:22] info: Temporalio.Worker.TemporalWorker[0]
      Worker cancelled, shutting down
[13:07:22] info: Temporalio.Worker.TemporalWorker[0]
      Worker cancelled, shutting down
[13:07:22] info: Temporalio.Worker.ActivityWorker[0]
      Beginning activity worker shutdown, will wait 00:00:00 before cancelling 1 activity instance(s)
[13:07:22] info: Temporalio.Worker.ActivityWorker[0]
      Beginning activity worker shutdown, will wait 00:00:00 before cancelling 0 activity instance(s)
[13:07:24] info: Temporalio.Activity:CleanupFileFromWorkerFileSystem[0]
      Removing C:\Users\jaken\AppData\Local\Temp\tmp92FC.tmp
[13:07:24] warn: Temporalio.Worker.TemporalWorker[0]
      Worker finalization failed
      System.InvalidOperationException: Cannot finalize, expected 1 reference, got 2
         at Temporalio.Bridge.Worker.FinalizeShutdownAsync()
         at Temporalio.Worker.TemporalWorker.ExecuteInternalAsync(Func`1 untilComplete, CancellationToken stoppingToken)
Unhandled ex System.ArgumentException: Something bad happened
   at Program.<>c.<<<Main>$>b__0_5>d.MoveNext() in C:\dev\temporalio\samples-dotnet\src\ActivityStickyQueues\Program.cs:line 55
--- End of stack trace from previous location ---
   at Temporalio.Worker.TemporalWorker.ExecuteInternalAsync(Func`1 untilComplete, CancellationToken stoppingToken)
   at Temporalio.Worker.TemporalWorker.ExecuteInternalAsync(Func`1 untilComplete, CancellationToken stoppingToken)
   at Program.<>c__DisplayClass0_0.<<<Main>$>g__RunWorkerAsync|0>d.MoveNext() in C:\dev\temporalio\samples-dotnet\src\ActivityStickyQueues\Program.cs:line 64
Unhandled exception. System.ArgumentException: Something bad happened
   at Program.<>c.<<<Main>$>b__0_5>d.MoveNext() in C:\dev\temporalio\samples-dotnet\src\ActivityStickyQueues\Program.cs:line 55
--- End of stack trace from previous location ---
   at Temporalio.Worker.TemporalWorker.ExecuteInternalAsync(Func`1 untilComplete, CancellationToken stoppingToken)
   at Temporalio.Worker.TemporalWorker.ExecuteInternalAsync(Func`1 untilComplete, CancellationToken stoppingToken)
   at Program.<>c__DisplayClass0_0.<<<Main>$>g__RunWorkerAsync|0>d.MoveNext() in C:\dev\temporalio\samples-dotnet\src\ActivityStickyQueues\Program.cs:line 64
--- End of stack trace from previous location ---
   at Program.<Main>$(String[] args) in C:\dev\temporalio\samples-dotnet\src\ActivityStickyQueues\Program.cs:line 89
   at Program.<Main>(String[] args)
```

Comment (cretz) [reply src/ActivityStickyQueues/Program.cs]: Yeah, that's just a warning and there may be a bug here on exception. We can file an issue, but I wouldn't worry about it for this sample.

File: src/ActivityStickyQueues/NonStickyActivities.cs [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+{
+    private readonly string uniqueWorkerTaskQueue;
+
+    public NonStickyActivities(string uniqueWorkerTaskQueue) =>
+        this.uniqueWorkerTaskQueue = uniqueWorkerTaskQueue;
+
+    [Activity]
+#pragma warning disable CA1024
+    public string GetUniqueTaskQueue() => uniqueWorkerTaskQueue;
+#pragma warning restore CA1024
```
[the remark this answers is not in the document]
Comment (cretz) [reply src/ActivityStickyQueues/NonStickyActivities.cs:15]: (actually, maybe not ...doesn't make much sense for activities most of the time)

**Review by cretz — APPROVED** [review]
Looks great! Will merge if/when CI passes.

**Pull request merged**